lib/SILGen/SILGenApply.cpp

```diff
@@ -1232,6 +1232,13 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
   }
 
   void visitFunctionConversionExpr(FunctionConversionExpr *e) {
+    // If this is a flattening function conversion, emit the expression
+    // directly.
+    if (e->isFlattening()) {
```

Comment:
What you should do is have a FlatteningExpr that wraps a DeclRefExpr. It would be handled just like a DeclRefExpr, except the underlying decl would be emitted with a different uncurry level.